src/components/Attachment/Attachment.tsx

```diff
@@ -0,0 +1,126 @@
+import * as _ from 'lodash'
+import * as PropTypes from 'prop-types'
+import * as React from 'react'
+
+import { UIComponent, customPropTypes, createShorthandFactory, createHTMLDivision } from '../../lib'
+import { Extendable, ItemShorthand } from '../../../types/utils'
+import { IComponentPartStylesInput, ComponentVariablesInput } from '../../../types/theme'
+import Icon from '../Icon/Icon'
+import Button from '../Button/Button'
+
+export type AttachmentProps = {
+  action?: ItemShorthand
+  actionable?: boolean
+  as?: any
+  children?: React.ReactChildren
+  content?: React.ReactNode
+  description?: string
+  header?: string
+  icon?: ItemShorthand
+  progress?: string | number
+  styles?: IComponentPartStylesInput
+  variables?: ComponentVariablesInput
+}
+
+/**
+ * An Attachment displays a file attachment.
+ */
+class Attachment extends UIComponent<Extendable<AttachmentProps>, any> {
+  static create: Function
+
+  static className = 'ui-attachment'
+
+  static displayName = 'Attachment'
+
+  static handledProps = [
+    'action',
+    'actionable',
+    'as',
+    'children',
+    'content',
+    'description',
+    'header',
+    'icon',
+    'progress',
+    'styles',
+    'variables',
+  ]
+
+  static propTypes = {
+    /** Button shorthand for the action slot. */
+    action: customPropTypes.itemShorthand,
+
+    /** An Attachment can be styled to indicate possible user interaction. */
+    actionable: PropTypes.arrayOf(customPropTypes.itemShorthand),
+
+    /** An element type to render as. */
+    as: customPropTypes.as,
+
+    /** Define your own children. */
+    children: PropTypes.node,
+
+    /** Shorthand for primary content. */
+    content: PropTypes.any,
+
+    /** A string describing the attachment. */
+    description: PropTypes.string,
+
+    /** The name of the attachment. */
+    header: PropTypes.string,
+
+    /** Shorthand for the icon. */
+    icon: customPropTypes.itemShorthand,
+
+    /** Value indicating percent complete. */
+    progress: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
+
+    /** Custom styles to be applied to the component. */
+    styles: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
+
+    /** Custom variables to be applied to the component. */
+    variables: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
+  }
+
+  renderComponent({ ElementType, classes, rest, styles, variables }) {
+    const { header, description, icon, action, progress } = this.props
+
+    return (
+      <ElementType {...rest} className={classes.root}>
+        <div className={classes.icon}>
+          {Icon.create(icon, {
+            defaultProps: {
+              size: 'big',
+              fitted: true,
+            },
+          })}
+        </div>
```

Comment:
I am wondering if this is the best approach, adding div's for handling our styling. I don't see a way how can the user alter these, unless using CSS selectors. I had similar propose https://github.com/stardust-ui/react/pull/199#issuecomment-419465588 but I don't think thins should be the general approach when defining shorthands (see the comments below). What are your thoughts on this @levithomason ?

Reply:
Yes, exactly. Not strictly for this component, but just as e reference to the RFC about the customizing of the slot's tree.